### PR TITLE
Laura MediaRanker

### PR DESCRIPTION
**Pull request by lauracodecreations**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I create the joining_date method to define the date by calling the created_at attribute, and in addition, I formatted the date.
Describe how you approached testing that model method. What edge cases did you come up with? | I plan to continue working on this. I did test my work model validations
What are `session` and `flash`? What is the difference between them? | session allows users to log in, flash is a message that flashes the user if the 
Describe a controller filter you wrote. | not yet
What was one thing that you gained more clarity on through this assignment? | how to debug errors in rails when the 'raise' does not work, how to create a custom method in the model, and how to test validations. 
What is the Heroku URL of your deployed application? | https://media-app-2018.herokuapp.com/users/1
Do you have any recommendations on how we could improve this project for the next cohort? | maybe learning how to use pry to test the params in the rails app while is running would be helpful for other students when debugging syntax errors. A tutor taught me how to use pry to stop the rails in a method and test if the syntax is correct (the webpage will not change move unless I exit pry). This helped a lot when debugging errors because it was more efficient than writing all info from params in the rails console.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Rails.application.routes.draw do
+  get 'users/index'
+  get 'users/create'
+  get 'users/new'
+  get 'users/show'
+  get 'users/destroy'
```
Comment (dHelmgren) [inline config/routes.rb:6]: I don't think you need all these routes for `users` - only `index` and `show` are needed for this project.

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+<h1>Media</h1>
+<h3>Albums</h3>
+<table>
+<tr><td>Votes</td><td>Title</td><td>Created By</td><td>Published</td><td>Upvote</td></tr>
+<% @album.each do |work| %>
+<tr>
+  <td><%=work.votes.length%></td>
+  <td><%= link_to "#{work.title}", work_path(work.id) %></td><td>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:8]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: app/models/vote.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Vote < ApplicationRecord
+  validates :user, uniqueness: { scope: :user_id,
+    message: "should happen once per user" }
```
Comment (dHelmgren) [inline app/models/vote.rb:3]: I think you misinterpreted the request we had. We wanted person to be able to vote for more than one work, but we didn't want users to be able to vote for _a single work_ more than once. 

Put another way, if I upvote DAMN by Kendrick Lamar, I can still upvote Titanic, but I can't vote for DAMN again.

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Work < ApplicationRecord
+  has_many :votes
+  CATEGORIES = %w(album book movie)
```
Comment (dHelmgren) [inline app/models/work.rb:3]: Nice job with this, you can actually leverage this EVEN MORE! Think about what this gives you on your works/index page!

File: test/models/vote_test.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require "test_helper"
+
+describe Vote do
+  let(:vote) { Vote.new }
+
+  it "must be valid" do
+    value(vote).must_be :valid?
+  end
```
Comment (dHelmgren) [inline test/models/vote_test.rb:8]: You should be testing the uniqueness constraint here! Specific test cases I'd want to see:

- A user can have votes for two different works
- A work can have votes from two different users
- A user cannot vote for the same work twice

File: test/models/work_test.rb [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    it 'has one of the allowed categories' do
+      work.category = "music"
+      expect(work).wont_be :valid?
+
+      Work::CATEGORIES.each do | category |
+        work.category = category
+        expect(work).must_equal :valid?
+      end
+    end
+
```
Comment (dHelmgren) [inline test/models/work_test.rb:123]: This is where I would expect tests for the code that picks your top 10 works to live.

For implementing those I would wonder:

- What happens if there are no works?
- What happens if there are works but no votes?
- What happens if two works have the same number of votes?

Similarly for you category and top ten methods, I would ask:

- What if there are no works of that category?
- What if there are less than 10 works?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | Good
Comprehension questions	| Yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Generally good, see comments
Views are well-organized (DRY, use of semantic HTML, use of partials) | Mostly true, but see comments!
Errors are reported to the user | Yes
Business logic lives in the models | Business logic is missing.
Models are thoroughly tested, including relations, validations and any custom logic | Tests could be more robust.
**Wave 1 - Media** |
Splash page shows the three media categories | No
Basic CRUD operations on media are present and functional | Mostly, but there are a few bugs, see comments.
**Wave 2 - Users and Votes** |
Users can log in and log out | Yes
The ID of the current user is stored in the session | Yes!
A user cannot vote for the same media more than once | Yes, but see comments
All media lists are ordered by vote count | No.
Splash page contains a media spotlight | No.
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | No.
Individual user pages and the user list are present | Yes
**Optional - Styling** |
Bootstrap is used appropriately | Yes!
Look and feel is similar to the original | No, but not important
**Overall** | The parts of the project that you were able to finish work very well and are cleverly designed, but it feels to me like you ran out of time. If there is anything I would suggest that you practice, it's writing business logic similar to the Top Ten code that was required for this project. Keep up the good work!
